core/trino-main/src/main/java/io/trino/execution/scheduler/TaskDescriptorStorage.java

```diff
@@ -0,0 +1,249 @@
+    /**
+     * Get task descriptor
+     *
+     * @return Non empty {@link TaskDescriptor} for a task identified by the <code>stageId</code> and <code>partitionId</code>.
+     * Returns {@link Optional#empty()} if the query of a given <code>stageId</code> has been finished (e.g.: cancelled by the user or finished early).
+     * @throws java.util.NoSuchElementException if {@link TaskDescriptor} for a given task does not exist
+     */
+    public synchronized Optional<TaskDescriptor> get(StageId stageId, int partitionId)
+    {
+        TaskDescriptors storage = storages.get(stageId.getQueryId());
+        if (storage == null) {
+            // query has been terminated
+            return Optional.empty();
+        }
+        return Optional.of(storage.get(stageId, partitionId));
+    }
+
+    /**
+     * Notifies the storage that the query with a given <code>queryId</code> has been finished and the task descriptors can be safely discarded.
+     * <p>
+     * The engine may decided to destroy the storage while the scheduling is still in process (for example if query was cancelled). Under such
+     * circumstances the implementation will ignore future calls to {@link #put(StageId, TaskDescriptor)} and return
+     * {@link Optional#empty()} from {@link #get(StageId, int)}. The scheduler is expected to handle this condition appropriately.
+     */
+    public synchronized void destroy(QueryId queryId)
+    {
+        TaskDescriptors storage = storages.remove(queryId);
+        if (storage != null) {
+            updateMemoryReservation(-storage.getReservedBytes());
+        }
+    }
+
+    private synchronized void updateMemoryReservation(long delta)
+    {
+        reservedBytes += delta;
+        while (reservedBytes > maxTaskDescriptorStorageMemoryInBytes) {
+            // drop a query that uses the most storage
+            QueryId queryId = storages.entrySet().stream()
+                    .max(Comparator.comparingLong(entry -> entry.getValue().getReservedBytes()))
+                    .map(Map.Entry::getKey)
+                    .orElseThrow(() -> new VerifyException(format("storage is empty but retainedSizeInBytes (%s) is still greater than totalStorageCapacityInBytes (%s)", reservedBytes, maxTaskDescriptorStorageMemoryInBytes)));
+            TaskDescriptors storage = storages.get(queryId);
+            long previousReservedBytes = storage.getReservedBytes();
+            storage.fail(new TrinoException(
+                    EXCEEDED_TASK_DESCRIPTOR_STORAGE_CAPACITY,
+                    format("Task descriptor storage capacity has been exceeded: %s > %s", succinctBytes(maxTaskDescriptorStorageMemoryInBytes), succinctBytes(reservedBytes))));
+            long currentReservedBytes = storage.getReservedBytes();
+            reservedBytes += (currentReservedBytes - previousReservedBytes);
+        }
+    }
+
+    @Managed
+    public synchronized long getReservedBytes()
+    {
+        return reservedBytes;
+    }
+
+    @NotThreadSafe
```

Comment:
nit: not important as this is internal class.

Reply:
I just tough it wouldn't hurt to declare the intention explicitly so then nobody is tempted to use this object outside the lock or try enforce synchronization in the object itself.